agent/checker.go

```diff
@@ -4,31 +4,61 @@ import (
 	"fmt"
 	"io"
 	"io/ioutil"
+	"log"
 	"strconv"
+	"strings"
 
 	"github.com/aquasecurity/bench-common/check"
 	"github.com/pkg/errors"
 )
 
 type Checker func() (CheckResult, error)
 
-func NewChecker(definitionsPath string) (Checker, error) {
-	data, err := ioutil.ReadFile(definitionsPath)
-	if err != nil {
-		return nil, errors.Wrap(err, "could not read definitions file")
+/* Remove
+  '---
+   controls:
+	version: ...
+	id: ...
+	description: ...
+	type: ...
+	groups:
+'
+that comes before '    - id:...'
+You need to do it to append the yaml to another yaml file
+*/
+func trim_yaml_header(yaml_text string) string {
+	pattern := "\ngroups:"
+	index := strings.Index(yaml_text, pattern)
+	if index == -1 {
+		log.Fatal("the yaml file doesn't contain the 'id:' field and is possibly broken")
 	}
+	index += len(pattern)
+	return yaml_text[index:]
+}
 
+func NewChecker(definitionsPaths []string) (Checker, error) {
+	var data [][]byte
+	for _, path := range definitionsPaths {
+		datum, err := ioutil.ReadFile(path)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not read definitions file")
+		}
+		data = append(data, datum)
+	}
 	return func() (CheckResult, error) {
-		result := CheckResult{}
-		controls, err := check.NewControls(data, nil)
+		var result CheckResult
+		first_yaml_file := string(data[0])
+		// Appent all next files to the first
+		for _, datum := range data[1:] {
+			first_yaml_file += trim_yaml_header(string(datum))
```

Comment:
interesting, so this is merging them textually into one yaml and then parse them in one go. what is the trade off over https://stackoverflow.com/questions/51947879/read-and-merge-two-yaml-files ?

is there a way to make the bench tool parse a list of strings/yamls instead of just one? I wasn't able to clearly find an answer on that in a cursory look at the code.

Reply:
AFAI understand the https://stackoverflow.com/questions/51947879/read-and-merge-two-yaml-files merges files so that the properties from the next file overwrite those from the former. We have many description: properties and we don't want them to be overwritten.

I was unsuccessfully trying to parse only a half of a yaml file. It's possible the unparsed part will simply be lost. Though I could do something that does somewhat, but let's decide on Thursday about the requirements for the expressiveness of our cheks.